### PR TITLE
Fixes: #1290. iPad will now have Desktop UA as default in iOS 13 with a setting to disable it

### LINKED ISSUE #1290: User agent on iPad should default to a desktop one.
### Steps to Reproduce 
  1. Visit a website

**Actual result:** 

**Expected result:**

**Reproduces how often:** [Easily reproduced, Intermittent Issue]

**Brave Version:** 

**Device details:** 

**Website problems only:**
- did you check with Brave Shields down?
- did you check in Safari/Firefox (WkWebView-based browsers)?

### PR DESCRIPTION
**Pull request by danishjafri88**
Fixes: #1290, iPad will use Desktop UA by default.

To test please visit https://www.whatsmyua.info
and load the page in iPad/iPhone (iOS 12 & 13) and also try changing the mode from share option

On iOS 12 (iPhone & Pad)& iOS 13 (iPhone), there should be no change and should be same as older versions,
while on iOS 13 iPad the launch should be. in desktop and the setting will be available 
## Submitter Checklist:

- [x] Submitted a [ticket](https://github.com/brave/brave-ios/issues) for my issue if one did not already exist.
- [x] My patch or PR title has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!` (or `No Bug: <message>` if no relevant ticket)
- [x] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`
- [x] New files have MPL-2.0 license header.

## Reviewer Checklist:

- [ ] PR is linked to an issue via [Zenhub](https://www.zenhub.com/extension).
- [ ] Issues are assigned to at least one epic.
- [ ] Issues include necessary QA labels:
  - [ ] `QA/(Yes|No)`
  - [ ] `release-notes/(include|exclude)`
  - [ ] `bug` / `enhancement`
- [ ] Necessary security reviews have taken place.
- [ ] Adequate test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable)

### CODE REVIEW

**Review by kylehickinson — COMMENTED** [review]
Do we want to use desktop UA even when the user is using Brave in split-screen/multitasking and the horizontal size class becomes `compact`?

**Comment by Brandon-T** [comment]
Running your code on Youtube still has this one problem from WebKit (also exists in Firefox iOS):

Screen Shot 2019-08-09 at 8 50 33 PM

We actually need BOTH of these PRs for iPad. Danish's PR sets the User-Agent to default.. and mine fixes the Full-Screen issue: https://github.com/brave/brave-ios/pull/1366

**Comment by Brandon-T** [comment]
For iOS 13:

```
if #available(iOS 13.0, *) {
    configuration.defaultWebpagePreferences.preferredContentMode = .recommended
}
```


We can use `.recommended`:

```swift
/*
WKContentModeRecommended -  The recommended content mode for the current platform
WKContentModeMobile  -     Represents content targeting mobile browsers
WKContentModeDesktop  -     Represents content targeting desktop browsers
WKContentModeRecommended - behaves like WKContentModeMobile on iPhone and iPad mini
     and WKContentModeDesktop on other iPad models as well as Mac.
*/
```

Still same FS problem though but that's about it. It'll take care of Desktop UA itself.

File: Client/Frontend/Browser/Tab.swift [hunk 1]
```diff
@@ -123,7 +123,7 @@ class Tab: NSObject {
 
     /// Whether or not the desktop site was requested with the last request, reload or navigation. Note that this property needs to
     /// be managed by the web view's navigation delegate.
-    var desktopSite: Bool = false
+    var desktopSite: Bool = UIDevice.isIpad ? true : false
```
Comment (iccub) [inline Client/Frontend/Browser/Tab.swift]: This should be a `Preference` that can be set in settings. Turned on by default on iPads.

I've started some initial work on it here if you want to take a look https://github.com/brave/brave-ios/compare/development...iccub:bugfix/1290?expand=1

**Review by iccub — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/Tab.swift [hunk 2]
```diff
@@ -123,7 +123,7 @@ class Tab: NSObject {
 
     /// Whether or not the desktop site was requested with the last request, reload or navigation. Note that this property needs to
     /// be managed by the web view's navigation delegate.
-    var desktopSite: Bool = false
+    var desktopSite: Bool = Preferences.General.alwaysRequestDesktopSite.value ? true : false
```
Comment (jhreis) [inline Client/Frontend/Browser/Tab.swift]: `Preferences.General.alwaysRequestDesktopSite.value` is a `Bool`, no need for the ternary operator here.

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 3]
```diff
@@ -1132,7 +1133,11 @@ class BrowserViewController: UIViewController {
     func restoreSpoofedUserAgentIfRequired(_ webView: WKWebView, newRequest: URLRequest) {
         // Restore any non-default UA from the request's header
         let ua = newRequest.value(forHTTPHeaderField: "User-Agent")
-        webView.customUserAgent = ua != UserAgent.defaultUserAgent() ? ua : nil
+        if #available(iOS 13.0, *) {
```
Comment (Brandon-T) [inline Client/Frontend/Browser/BrowserViewController.swift]: We use if-available here but above we is `islessthan13`

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 4]
```diff
@@ -1277,8 +1282,16 @@ class BrowserViewController: UIViewController {
             }
         }
         
-        // Remember whether or not a desktop site was requested
-        tab.desktopSite = webView.customUserAgent?.isEmpty == false
+        // Remember whether or not a desktop/mobile site was requested
+        if #available(iOS 13.0, *) {
+            if let customUA = webView.customUserAgent, customUA.isEmpty == false {
```
Comment (Brandon-T) [inline Client/Frontend/Browser/BrowserViewController.swift]: `customUA.isEmpty == false` can be changed to `!customUA.isEmpty`

File: Client/Frontend/Browser/Tab.swift (hunk 2 above)
Comment (Brandon-T) [inline Client/Frontend/Browser/Tab.swift]: No need for branch statement `? true : false` because the `alwaysRequestDesktopSite.value` already results in a boolean.

I think we can just do:

`var desktopSite = Preferences.General.alwaysRequestDesktopSite.value`

File: Client/Frontend/Browser/TabManager.swift [hunk 5]
```diff
@@ -165,6 +165,11 @@ class TabManager: NSObject {
     private class func getNewConfiguration() -> WKWebViewConfiguration {
         let configuration = WKWebViewConfiguration()
         configuration.processPool = WKProcessPool()
+        // Setting the Unique applicationName for the user agent to identify Brave.
+        if #available(iOS 13.0, *) {
+            configuration.applicationNameForUserAgent = Preferences.General.alwaysRequestDesktopSite.value ?
+            UserAgent.desktopUAApplicationName() : UserAgent.mobileUAApplicationName()
```
Comment (Brandon-T) [inline Client/Frontend/Browser/TabManager.swift:171]: In both cases won't this value be: `FxiOS`?

File: Shared/SwiftAttributes.swift [hunk 6]
```diff
@@ -0,0 +1,12 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
+
+public var lessThaniOS13: Bool {
```
Comment (Brandon-T) [inline Shared/SwiftAttributes.swift:7]: Used only in one place, can we make it a lambda/location function for that one scenario instead of a global variable?

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift [hunk 7]
```diff
@@ -64,6 +64,16 @@ extension BrowserViewController: WKNavigationDelegate {
         return false
     }
 
+    // This is the iOS 13 delegate callback which is similar to the delegate below it
+    // except that a WKWebpagePreferences is provided, Setting the type of contentMode loads the site accordingly.
+    @available(iOS 13.0, *)
+    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, preferences: WKWebpagePreferences, decisionHandler: @escaping (WKNavigationActionPolicy, WKWebpagePreferences) -> Void) {
+        self.webView(webView, decidePolicyFor: navigationAction) {
+            preferences.preferredContentMode = Preferences.General.alwaysRequestDesktopSite.value ? .desktop :. mobile
+            decisionHandler($0, preferences)
```
Comment (Brandon-T) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:73]: This breaks `requestMobileSite` from the Share Menu because it only obeys your "settings" from `Preferences` and not your site's preference..

IE: Have AlwaysRequestDesktop site on.. Then go to youtube and go to share menu and choose `RequestMobileSite`, it will still show Desktop, but if you go back to that same share menu, it will show `RequestDesktopSite` even though you're already on desktop UA..

In other words, the text will change for the share menu button but it does nothing in the end.

**Review by Brandon-T — CHANGES_REQUESTED** [review]

File: Client/Frontend/Browser/TabManager.swift (hunk 5 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/TabManager.swift:171]: No, in mobile it will be FxiOS and for desktop it will be Safari

File: Client/Frontend/Browser/Tab.swift (hunk 2 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/Tab.swift]: Lol, bad copy

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift (hunk 7 above)
Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:73]: Its not because of the settings but somehow YouTube is setting the url param as app=desktop. If you instead try  https://www.whatsmyua.info or anyother site it shows it works fine.

Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:73]: let me see if I can fix it on our end

Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:73]: I found that the reason is because we are trying to open a mobile site ie `m.youtube.com` in Desktop. Youtube then redirects it to `https://www.youtube.com/?app=desktop`. When we change the UA with the param app=desktop no change is done. 

I propose that we also change our favourites URL from mobile to desktop in this case? @Brandon-T  @iccub @jhreis

Comment (danishjafri88) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:73]: It happens on Desktop as well

**Review by Brandon-T — APPROVED** [review]

**Review by iccub — CHANGES_REQUESTED** [review]
Please rebase to build with iOS 13

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift (hunk 7 above)
Comment (iccub) [reply Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:73]: It doesn't work for me.
I open https://www.whatsmyua.info, request mobile site, and doesn't switch to mobile

I agree we can remove `mobile` part from favorites, youtube and twitter

File: Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift [hunk 8]
```diff
@@ -64,6 +64,16 @@ extension BrowserViewController: WKNavigationDelegate {
         return false
     }
 
+    // This is the iOS 13 delegate callback which is similar to the delegate below it
+    // except that a WKWebpagePreferences is provided, Setting the type of contentMode loads the site accordingly.
+    @available(iOS 13.0, *)
+    func webView(_ webView: WKWebView, decidePolicyFor navigationAction: WKNavigationAction, preferences: WKWebpagePreferences, decisionHandler: @escaping (WKNavigationActionPolicy, WKWebpagePreferences) -> Void) {
+        self.webView(webView, decidePolicyFor: navigationAction) {
+            preferences.preferredContentMode = Preferences.General.alwaysRequestDesktopSite.value ? .desktop :. mobile
```
Comment (iccub) [inline Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift:72]: `.desktop :. mobile` add space after `:`

File: ClientTests/ClientTests.swift [hunk 9]
```diff
@@ -14,13 +14,12 @@ import Alamofire
 class ClientTests: XCTestCase {
 
     // Simple test to make sure the WKWebView UA matches the expected FxiOS pattern.
-    func testUserAgent() {
+    func testMobileUserAgent() {
```
Comment (iccub) [inline ClientTests/ClientTests.swift:17]: This test doesn't work for me, breaks on CI too

File: Shared/UserAgent.swift [hunk 10]
```diff
@@ -45,59 +45,51 @@ open class UserAgent {
                 return firefoxUA
             }
         }
-
         return nil
     }
 
+    private static var systemDefaultUA: String = ""
```
Comment (iccub) [inline Shared/UserAgent.swift:51]: `String` can be removed

**Review by iccub — CHANGES_REQUESTED** [review]
One unit test is failing

`Request mobile` site doesn't seem to be working for me(XCode beta 6)

File: BraveShared/BraveStrings.swift [hunk 11]
```diff
@@ -393,6 +393,7 @@ extension Strings {
     public static let Privacy = NSLocalizedString("Privacy", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Privacy", comment: "Settings privacy section title")
     public static let Security = NSLocalizedString("Security", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Security", comment: "Settings security section title")
     public static let Save_Logins = NSLocalizedString("SaveLogins", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Save Logins", comment: "Setting to enable the built-in password manager")
+    public static let AlwaysRequestDesktopSite = NSLocalizedString("AlwaysRequestDesktopSite", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Always request desktop site", comment: "Setting to always request desktop site")
```
Comment (jhreis) [inline BraveShared/BraveStrings.swift:396]: > Setting to always use the desktop version of a site

```suggestion
    public static let AlwaysRequestDesktopSite = NSLocalizedString("AlwaysRequestDesktopSite", tableName: "BraveShared", bundle: Bundle.braveShared, value: "Always request desktop site", comment: "Setting to always use the desktop version of a site")
```

File: Client/Frontend/Browser/BrowserViewController.swift [hunk 12]
```diff
@@ -3054,6 +3067,15 @@ extension BrowserViewController: PreferencesObserver {
         switch key {
         case Preferences.General.tabBarVisibility.key:
             updateTabsBarVisibility()
+        case Preferences.General.alwaysRequestDesktopSite.key:
+            tabManager.reset()
+            self.tabManager.reloadSelectedTab()
+            for tab in self.tabManager.allTabs where tab != self.tabManager.selectedTab {
```
Comment (jhreis) [inline Client/Frontend/Browser/BrowserViewController.swift:3073]: I think this is overkill. I think `tabManager.reset()` should be enough. We definitely _do not) want to reload _all_ tabs, this could create hundreds of webviews.

File: Client/Frontend/Browser/Tab.swift [hunk 13]
```diff
@@ -195,6 +195,11 @@ class Tab: NSObject {
             configuration!.preferences = WKPreferences()
             configuration!.preferences.javaScriptCanOpenWindowsAutomatically = false
             configuration!.allowsInlineMediaPlayback = true
+            if #available(iOS 13.0, *) {
+                configuration!.defaultWebpagePreferences = WKWebpagePreferences().then {
+                    $0.preferredContentMode = Preferences.General.alwaysRequestDesktopSite.value ? .desktop : .mobile
```
Comment (jhreis) [inline Client/Frontend/Browser/Tab.swift:200]: Not a big fan of the duplication:

https://github.com/brave/brave-ios/pull/1368/files#diff-c0ee444fbdc7d184b084fb4cc3e534fcR72

But may not easy way around this?

File: Client/Frontend/Browser/Tab.swift [hunk 14]
```diff
@@ -399,7 +404,16 @@ class Tab: NSObject {
     }
 
     func reload() {
-        let userAgent: String? = desktopSite ? UserAgent.desktopUserAgent() : nil
+        
+        var userAgent: String?
```
Comment (jhreis) [inline Client/Frontend/Browser/Tab.swift:408]: Rather than specifying the `userAgent` here, we should be specifying the `mobileUA`:

```
var mobileUA: String? = nil

if #available(iOS 13.0, *) {
    mobileUA = UserAgent.mobileUserAgent()
}

let userAgent: String? = desktopSite ? UserAgent.desktopUserAgent() : mobileUA
```

This is _much_ cleaner IMO.

@danishjafri88 when seeing code duplication like this (`userAgent = desktopSite ? UserAgent.desktopUserAgent() : `), it is best to try to identify how we can abstract the duplicated portions.

File: Client/Frontend/Settings/SettingsViewController.swift [hunk 15]
```diff
@@ -166,7 +166,10 @@ class SettingsViewController: TableViewController {
             }
             general.rows.append(row)
         }
-        
+        if #available(iOS 13.0, *), UIDevice.isIpad {
```
Comment (jhreis) [inline Client/Frontend/Settings/SettingsViewController.swift]: Let's remove this iPad check.

File: Shared/SwiftAttributes.swift (hunk 6 above)
Comment (jhreis) [inline Shared/SwiftAttributes.swift:7]: Let's just fully delete this.

File: Shared/UserAgent.swift [hunk 16]
```diff
@@ -45,59 +45,51 @@ open class UserAgent {
...
     public static func defaultUserAgent() -> String {
         assert(Thread.current.isMainThread, "This method must be called on the main thread.")
-
         if let firefoxUA = UserAgent.cachedUserAgent(checkiOSVersion: true) {
             return firefoxUA
         }
-
-        let webView = UIWebView()
-
+        // Reset UA
```
Comment (jhreis) [inline Shared/UserAgent.swift:61]: All of these changes look a bit suspicious. We should pull down FF's latest `Shared/UserAgent.swift` code, and use that. If that is what this is, great, if not, let's get rid of it.

File: Client/Frontend/Browser/TabManager.swift (hunk 5 above)
Comment (jhreis) [reply Client/Frontend/Browser/TabManager.swift:171]: This does not seem correct to me. Why is this? Is it something from FF upstream?

**Review by jhreis — CHANGES_REQUESTED** [review]

File: Shared/UserAgentExtension.swift [hunk 17]
```diff
@@ -0,0 +1,35 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+import Foundation
```
Comment (jhreis) [inline Shared/UserAgentExtension.swift:5]: Ditto: https://github.com/brave/brave-ios/pull/1368/files#r316382234

**Comment by jhreis** [comment]
Closing in favor of: https://github.com/brave/brave-ios/pull/1425
